### PR TITLE
[WIP] Implement implicit converter for F# optionals

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
As discussed in #1254, this PR attempts to improve C# consumption of F# optional parameters in two ways:
* ~~Automatically append the [OptionalAttribute](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.optionalattribute(v=vs.110).aspx) and [DefaultParameterValueAttribute](https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.defaultparametervalueattribute(v=vs.100).aspx) in all optional parameters of F# methods. These will be added in addition to the the already automatically appended [FSharp.Core.OptionalArgumentAttribute](https://msdn.microsoft.com/visualfsharpdocs/conceptual/core.optionalargumentattribute-class-%5bfsharp%5d).~~
* Implement an implicit converter `'T -> 'T option` in FSharp.Core.

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
Hmm, appveyor is failing to build. Reason seems unrelated to changes introduced here.

**Comment by KevinRansom** [comment]
@eiriktsarpalis

**Comment by KevinRansom** [comment]
@eiriktsarpalis 

On my machine I synched master, and build net40 proto coreclr notests.
Everything built okay.

I merged your changes and got these errors:  

````

       "d:\microsoft\visualfsharp\build-everything.proj" (default target) (1) ->
       "d:\microsoft\visualfsharp\src\fsharp-library-build.proj" (Build target) (2) ->
       "d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj" (Build target) (3) ->
       (CoreCompile target) ->
         d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\reflect.fsi(291,19): error FS3188: Type inference cau
       sed an inference type variable to escape its scope. Consider adding type annotations to make your code l
       ess generic. [d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
         d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\reflect.fsi(291,19): error FS3188: Type inference cau
       sed an inference type variable to escape its scope. Consider adding type annotations to make your code l
       ess generic. [d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\FSharp.Core.fsproj]
         d:\microsoft\visualfsharp\src\fsharp\FSharp.Core\reflect.fsi(291,19): error FS3188: Type inference 
   ......
````

Reverted your changes and rebuilt ...  everything built fine.  So I think the issue is due to your change.

Kevin

**Comment by eiriktsarpalis** [comment]
@KevinRansom thanks, seeing that too on my machine. Wonder what broke here..

**Comment by eiriktsarpalis** [comment]
For lack of time I've removed the auto-insertion implementation and simply kept the implicit converter. What do you think? /cc @dsyme

**Comment by dsyme** [comment]
@eiriktsarpalis This looks like a fine addition.   Update to master and add some test cases? 

I'll leave it to @KevinRansom and @otawfik-ms to decide if it fits on the F~ 4.1 schedule.

**Comment by eiriktsarpalis** [comment]
I'll add those asap. It'd be an immensely useful addition to F# 4.1

**Comment by eiriktsarpalis** [comment]
The tests would probably need to be written in C#. Should I add a C# tests
project?

**Comment by dsyme** [comment]
@eiriktsarpalis You can probably add tests to "tests/fsharp/core/fsfromfsviacs" (F# from F# via C#).  Put the tests in the C# code then trigger them from F# main

**Comment by cartermp** [comment]
:+1: on this making it into F# 4.1, even though we're feature complete.  This will be an immediate benefit to a lot of people who are doing plenty of F#/C# interop.  I think we should make an exception here.

**Comment by KevinRansom** [comment]
@cartermp @dsyme For now everything is headed for 4.1.  We are feature complete on planned features.  However ... opportunistic safe new features are still welcome, subject to concerns about compatability, risk etc ...


Kevin

File: tests/fsharp/core/fsfromfsviacs/lib2.cs [hunk 1]
```diff
@@ -60,3 +60,14 @@ internal class SomeClass
...
     }
 }
+
+namespace FSharpOptionalTests
+{
+    public class ApiWrapper 
+    {
+        public static Tuple<FSharpOption<T>, FSharpOption<int>> MethodWithOptionalParams<T>(T value1, int value2)
+        {
+            return OptionalParameterTests.API.MethodWithOptionalParams<T>(value1 = value1, value2 = value2);
```
Comment (dsyme) [inline tests/fsharp/core/fsfromfsviacs/lib2.cs]: Please invoke this method from the actual test.fs?  

Also please add further tests for other uses of the implicit conversion (ones that just utilize the conversion independent of optional parameters).

**Comment by dsyme** [comment]
Off topic now, but IIRC one issue I had with this:

> Automatically append the OptionalAttribute and DefaultParameterValueAttribute in all optional parameters of F# methods. 

was that OptionalAttribute wasn't available in all portable profiles (e.g. 259).  That meant that compilation would change as you move from profile 259 to profile 7 and .NET Standard, which would be unfortunate.

**Comment by eiriktsarpalis** [comment]
I think it's ready, tests are passing.

**Comment by forki** [comment]
2,986 additions, 0 deletions in surfacearea for silverlight ?

**Comment by dsyme** [comment]
@eiriktsarpalis That file can be deleted and the test disabled

**Comment by forki** [comment]
Much better. Thanks

**Comment by dsyme** [comment]
Just to mention that over skype I had asked Eirik to add more tests for interop, especially around null, just to check this doesn't break anything for C# consumers.

**Comment by eiriktsarpalis** [comment]
@dsyme I'll create a second PR that adds those.
